chunk/chunk_store.go

```diff
@@ -243,49 +243,104 @@ outer:
 	return filteredChunks, nil
 }
 
-func (c *Store) lookupMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
-	metricName, matchers, err := util.ExtractMetricNameFromMatchers(matchers)
+func (c *Store) lookupChunksByMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
+	userID, err := user.Extract(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	userID, err := user.Extract(ctx)
-	if err != nil {
-		return nil, err
+	// Get metric name from matchers
+	metricName, matchers, ok := util.ExtractMetricNameFromMatchers(matchers)
+
+	// If there is a metric name, return chunks using metric name
+	if ok {
+		return c.lookupChunksByMetricName(ctx, from, through, matchers, userID, metricName)
+	}
+
+	// If there is no metric name, we want return chunks for all metric names
+	metricNameQueries, err := c.schema.GetReadQueries(from, through, userID)
+	metricNameEntries, err := c.lookupEntriesByQueries(ctx, metricNameQueries)
+
+	incomingChunkSets := make(chan ByKey)
+	incomingErrors := make(chan error)
+	for _, metricNameEntry := range metricNameEntries {
+		go func(metricName model.LabelValue) {
+			chunks, err := c.lookupChunksByMetricName(ctx, from, through, matchers, userID, metricName)
+			if err != nil {
+				incomingErrors <- err
+			} else {
+				incomingChunkSets <- chunks
+			}
+		}(model.LabelValue(metricNameEntry.Value))
 	}
 
+	var chunkSets []ByKey
+	var lastErr error
+	for i := 0; i < len(metricNameEntries); i++ {
+		select {
+		case incoming := <-incomingChunkSets:
+			chunkSets = append(chunkSets, incoming)
+		case err := <-incomingErrors:
+			lastErr = err
+		}
+	}
+
+	// Merge chunkSets in order
+	return nWayIntersect(chunkSets), lastErr
```

Comment:
In this case we don't want to interest, but rather union - when doing a query like `{__name__=~"foo.*"}`, we want to return chunks for the both series `foobar` and `foobaz`, not the the intersection of `foobar` and `foobar`  (which will be empty).